File: data-structures/graphs/undirected-graphs/undirectedGraph.js
/**
 * - My implementation of an undirected graph as a JS class
 * - Overall space complexity: O(V + E)
 */
class UndirectedGraph {
	/**
	 * - Instantiates a new undirected graph 
	 * - Time complexity: O(1)
	 * - Space complexity: O(1)
	 */
	constructor() {
		this.vertexCount = 0;
		this.adjacencyList = new Map();
	}

	/**
	 * - Adds a new vertex to the graph
	 * - Time complexity: O(1)
	 * - Space complexity: O(1)
	 */
	addVertex(vertex) {
		if (this.adjacencyList.has(vertex)) {
			console.log(`Graph already contains a vertex with value of ${vertex}`);
			return false;
		}

		this.adjacencyList.set(vertex, []);
		this.vertexCount++;
	}

	/**
	 * - Adds a new undirected edge between 2 nodes in the graph
	 * - Time complexity: O(V)
	 * - Space complexity: O(1)
	 */
	addEdge(vertex1, vertex2) {
		if (!(this.adjacencyList.has(vertex1) && this.adjacencyList.has(vertex2))) {
			console.log('Please ensure both vertices have been added to the graph before attempting to add an edge between them');
			return false;
		} else if (this.adjacencyList.get(vertex1).includes(vertex2)) { // O(V) check
			console.log(`Edge already exists between vertices ${vertex1} and ${vertex2}`);
			return false;
		}

		this.adjacencyList.get(vertex1).push(vertex2);
		this.adjacencyList.get(vertex2).push(vertex1);
	}

	/**
	 * - Print all the existing edges in the graph
	 * - Time complexity: O(1)
	 * - Space complexity: O(1)
	 */
	showConnections() {
		console.log(this.adjacencyList);
	}
}

File: data-structures/graphs/undirected-graphs/undirected_graph.py
class UndirectedGraph:
	"""
	My implementation of an undirected graph as a Python class

	Overall space complexity: O(V + E)
	"""
	def __init__(self):
		"""
		Instantiates a new undirected graph

		Time complexity: O(1)

		Space complexity: O(1)
		"""

		self.vertex_count = 0
		self.adjacency_list = {}

	def add_vertex(self, vertex):
		"""
		Adds a new vertex to the graph

		Time complexity: O(1)

		Space complexity: O(1)
		"""
		assert (vertex not in self.adjacency_list), f"Graph already contains a vertex with value of {vertex}"

		self.adjacency_list[vertex] = []
		self.vertex_count += 1

	def add_edge(self, vertex_1, vertex_2):
		"""
		Adds a new undirected edge between 2 nodes in the graph

		Time complexity: O(V)

		Space complexity: O(1)
		"""
		assert (vertex_1 in self.adjacency_list and vertex_2 in self.adjacency_list), "Please ensure both vertices have been added to the graph before attempting to add an edge between them"
		# O(V) operation to check if edge already exists
		assert (vertex_2 not in self.adjacency_list[vertex_1]), f"Edge already exists between vertices {vertex_1} and {vertex_2}"

		self.adjacency_list[vertex_1].append(vertex_2)
		self.adjacency_list[vertex_2].append(vertex_1)

	def get_adjacent_vertices(self, vertex):
		"""
		Return all adjacent vertices of a given vertex in the graph

		Time complexity: O(1)

		Space complexity: O(1)
		"""
		assert vertex in self.adjacency_list, f"Graph does not contain a vertex with value of {vertex}"

		return self.adjacency_list[vertex]

	def show_connections(self):
		"""
		Print all existing edges in the graph

		Time complexity: O(1)

		Space complexity: O(1)
		"""

		print(self.adjacency_list)